### PR TITLE
fix: Broken user group add and removal in user app

### PR DESCRIPTION
**Pull request by netroms**
### Summary

The user app was broken and did not work when trying to add or remove a user group on the user.
This was broken because the new patch method in the AbstractCrudController, that now the UI uses, don't support inverted relationship operations. 
To fix this, the relationship had be inverted and let the UserGroup service handle the operations, by using the pre and post patch hooks.

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/collection/CollectionUtils.java [hunk 1]
```diff
@@ -291,4 +293,16 @@ public static <K, V> Map<K, V> mapOf(
...
+    /**
+     * Convert an Iterator to a Stream.
+     *
+     * @param iterator The Iterator to convert to a stream.
+     * @return A stream of the iterable.
+     */
+    public static <E> Stream<E> iterableToStream( Iterator<E> iterator )
+    {
+        Iterable<E> iterable = () -> iterator;
+        return StreamSupport.stream( iterable.spliterator(), false );
```
Comment (jbee) [inline dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/collection/CollectionUtils.java:306]: Better use : 
```java
StreamSupport.stream(spliteratorUnknownSize(iter, Spliterator.ORDERED | Spliterator.NONNULL | Spliterator.IMMUTABLE), false);
```
`Spliterator.NONNULL` might not apply.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java [hunk 2]
```diff
@@ -816,13 +823,31 @@ protected void prePatchEntity( User oldEntity, User newEntity )
...
         {
-            userService.expireActiveSessions( user );
+            userService.expireActiveSessions( entityAfter );
+        }
+
+        if ( entityAfter != null && patch != null )
+        {
+            for ( JsonPatchOperation op : patch.getOperations() )
+            {
+                JsonPointer userGroups = op.getPath().matchProperty( "userGroups" );
```
Comment (mortenoh) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java:839]: I wonder if its better to match on `/userGroups` as we just want this supported at the root I guess?

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java [hunk 3]
```diff
@@ -816,13 +823,31 @@ protected void prePatchEntity( User oldEntity, User newEntity )
...
+                String opName = op.getOp();
+                if ( userGroups != null && opName.equals( "add" ) )
```
Comment (mortenoh) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java:841]: Check for `replace` also, as in this case they have the same semantics

**Review by mortenoh — CHANGES_REQUESTED** [review]

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java (hunk 2 above)
Comment (netroms) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java:839]: Yeah, that's a valid point. I will change.

**Review by jbee — APPROVED** [review]
We should think about how we can solve situation like this without adding workarounds.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java (hunk 3 above)
Comment (netroms) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java:841]: ok

File: dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/collection/CollectionUtils.java (hunk 1 above)
Comment (netroms) [reply dhis-2/dhis-support/dhis-support-commons/src/main/java/org/hisp/dhis/commons/collection/CollectionUtils.java:306]: Ok

**Review by mortenoh — APPROVED** [review]

**Pull request merged**